sys/net/gnrc/link_layer/gomach/gomach_internal.c

```diff
@@ -0,0 +1,1335 @@
+        return -ENOBUFS;
+    }
+    gomach_pkt = pkt;
+
+    gnrc_pktsnip_t *netif_snip = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_NETIF);
+    if (netif_snip == NULL) {
+        LOG_ERROR("[GOMACH]: NO netif_hdr found in gnrc_gomach_send_preamble_ack().\n");
+        gnrc_pktbuf_release(gomach_pkt);
+        return -ENOBUFS;
+    }
+    else {
+        nethdr_preamble_ack = netif_snip->data;
+    }
+
+    /* Construct NETIF header and insert address for preamble-ACK packet. */
+    gnrc_netif_hdr_init(nethdr_preamble_ack, 0, 0);
+
+    /* Send preamble-ACK as broadcast. */
+    nethdr_preamble_ack->flags |= GNRC_NETIF_HDR_FLAGS_BROADCAST;
+
+    int res = gnrc_gomach_send(netif, pkt, NETOPT_DISABLE);
+    if (res < 0) {
+        LOG_ERROR("ERROR: [GOMACH]: send preamble-ack failed in gnrc_gomach_send_preamble_ack().\n");
+        gnrc_pktbuf_release(gomach_pkt);
+    }
+    return res;
+}
+
+int gnrc_gomach_send_beacon(gnrc_netif_t *netif)
+{
+    assert(netif != NULL);
+
+    uint8_t i;
+    uint8_t j = 0;
+    uint8_t total_tdma_node_num = 0;
+    uint8_t total_tdma_slot_num = 0;
+    gnrc_pktsnip_t *pkt = NULL;
+    gnrc_pktsnip_t *gomach_pkt = NULL;
+    gnrc_netif_hdr_t *nethdr_beacon = NULL;
+
+    /* Start assemble the beacon packet */
+    gnrc_gomach_frame_beacon_t gomach_beaocn_hdr;
+    gomach_beaocn_hdr.header.type = GNRC_GOMACH_FRAME_BEACON;
+    gomach_beaocn_hdr.sub_channel_seq = netif->mac.prot.gomach.sub_channel_seq;
+
+    /* Start generating the slots list and the related ID list for guiding
+     * the following vTMDA procedure (slotted transmission). */
+    netif->mac.rx.vtdma_manag.total_slots_num = 0;
+
+    gnrc_gomach_l2_id_t id_list[GNRC_GOMACH_SLOSCH_UNIT_COUNT];
+    uint8_t slots_list[GNRC_GOMACH_SLOSCH_UNIT_COUNT];
+
+    for (i = 0; i < GNRC_GOMACH_SLOSCH_UNIT_COUNT; i++) {
+        if (netif->mac.rx.slosch_list[i].queue_indicator > 0) {
+            /* Record the device's (that will be allocated slots) address to the ID list. */
+            memcpy(id_list[j].addr,
+                   netif->mac.rx.slosch_list[i].node_addr.addr,
+                   netif->mac.rx.slosch_list[i].node_addr.len);
+
+            /* Record the number of allocated slots to the slots list. */
+            slots_list[j] = netif->mac.rx.slosch_list[i].queue_indicator;
+
+            total_tdma_node_num++;
+            total_tdma_slot_num += slots_list[j];
+
+            /* If there is no room for allocating more slots, stop. */
+            if (total_tdma_slot_num >= GNRC_GOMACH_MAX_ALLOC_SLOTS_NUM) {
+                uint8_t redueced_slots_num;
+                redueced_slots_num = total_tdma_slot_num - GNRC_GOMACH_MAX_ALLOC_SLOTS_NUM;
+                slots_list[j] -= redueced_slots_num;
+                total_tdma_slot_num -= redueced_slots_num;
+                break;
+            }
+            j++;
+        }
+    }
+
+    gomach_beaocn_hdr.schedulelist_size = total_tdma_node_num;
+
+    if (total_tdma_node_num > 0) {
```

Comment:
This if-block could also maybe be its own function.

Reply:
Addressed~